### ISSUE TITLE
VAX idling on VMS 5.0 & 5.1

### ISSUE THREAD

**Issue by ethump** [simh/simh#853]
Hi,

Hopefully this is the right forum. I've noticed that when my simulated VAXes are running either VAX/VMS 5.0 or 5.1 the simh process slurps 100% CPU.

All other versions I've tried 4.6, 4.7 and 5.2 and above all work as expected. set cpu idle=vms set in all cases.

Tried MicroVAX II and MicroVAX 3900 simulators on both Windows and Linux hosts. All combinations exhibit same behavior.

Simulator version 4.0-0, commit id: cabd3784.

Happy to share images etc. SIMH config is very basic, just the idle=vms and disk attach commands.

If this is not the appropriate forum, grateful if someone could point me at the right place?

Cheers all!

E

**Comment by markpizz** [comment]
Hi @ethump

Please put your  full name in your GitHub profile so we know who we're talking to.

Meanwhile, I was pretty sure that, although the VMS idling code changed when VMS went from 4.x to 5.0 (with the introduction of SMP support), it remained stable after that.  I'll be glad to be proven wrong, so if you can supply a minimal disk image that demonstrates this, I'll track down what is going on and add support.

I suspect that Standalone Backup running from an empty disk would represent a minimal disk image that demonstrates this.  You could generate such a thing in a small container with the following:
```
sim> set rq3 RD53
sim> set rq3 FORMAT=VHD
sim> attach rq3 STABACKUP.vhd
sim> attach rq0 your-normal-VMSv5.0.dsk
```
Once VMS is booted, login as SYSTEM and create the standalone backup disk with:
```
$ @SYS$UPDATE:STABACKIT DUA3:
```
Answer any prompts with defaults.
Once that completes shutdown the system and exit the simulator
You should have a relatively small STABACKUP.vhd container file.  
Please boot from that disk to confirm that idling isn't behaving as expected.
If it fails, please zip that container file and drop it here in this issue.

**Comment by ethump** [comment]
Hi, thanks for the follow up and thanks also for the awesome work on this project!
I'll sort my profile out, I just created it quickly to send this message. 
Please find STABACKUP attached per your direction, and I can confirm that when I boot it, it displays the behaviour described, ie 100% processor utilisation.
[STABACKUP.vhd.gz](https://github.com/simh/simh/files/4641248/STABACKUP.vhd.gz)

**Comment by ethump** [comment]
Please find attached also same from 5.1 STABACKUP
cheers!
[STABACKUP51.vhd.gz](https://github.com/simh/simh/files/4641273/STABACKUP51.vhd.gz)

**Comment by ethump** [comment]
Hi... I pulled the latest and compiled. Seems to work.... my 5.0 and 5.1 instances are idling at the same CPU usage as the rest!

Awesome stuff!!

Can I ask what (in simple terms!) the difference is between the VMS versions?

cheers
ed

**Comment by markpizz** [comment]
> Can I ask what (in simple terms!) the difference is between the VMS versions?

Well, the VAX computer architecture/instruction set has no specific "I've got nothing to do, so wait for an interrupt" - instruction (as opposed to others like the PDP-11 which has a specific WAIT instruction which serves this purpose.  So, the concept of Idle in VMS (and any other operating systems that run on a VAX) is implemented conceptually in the kernel as some sort of loop, which looks for 'something' that changes after an interrupt.  As you might imagine, this concept of checking for 'something' is somewhat vague and will be completely dependent on how the authors of whatever version of which ever operating system is running did things.  Over the life of an operating system the OS development team may change things which affect this and thus modify the idle checking code sequence.  The earliest versions of VMS actually had a process which had the lowest possible scheduling priority, so when nothing else was runnable, that idle process was merely scheduled.  It contained a single instruction which was a branch to the same branch instruction.  When interrupts occurred on that system, they might be due to an I/O completion that some process was waiting for, so that process would then be scheduled and the idle process would be preempted.  That original idle concept and implementation (branch to self instruction) existed until VMS V5 when Symmetric Multi Processing (SMP) support was added to the OS.  Under these conditions, different criteria for scheduling and idling exist and the idle code changed.  When I added the initial idle support, I had VMS V5.5-2 and V7.3 at my fingertips.  As I recall these both had different code sequences during the 'idle loop', but there was a common instruction that both used uniquely which didn't appear in other operating systems.  The simh trick to adding idle support for a given OS is merely to turn on CPU instruction history and run the OS until it is clearly 'idle' (after a boot finishes, or when it gets to the >>> prompt).  You then interrupt the simulator and examine the instruction execution history.  You look for a unique set of instructions and instruction conditions.  Conditions include things like running in system space, and/or at a particular IPL and/or on the interrupt stack, etc.  Look in VAX/vax_cpu.c for calls to cpu_idle().
